tests/_internal/test_resolve_target.py

```diff
@@ -9,81 +9,101 @@
+        read_online_id.return_value = online_id
         candidate = CandidateDeviceFactory()
 
-        subject = _build_target_resolver(candidate)()
+        subject = resolve_target(candidate)
 
-        self.assertEqual(subject, get_target_by_product_code.return_value)
-        get_target_by_product_code.assert_called_once_with(read_product_code.return_value)
-        read_product_code.assert_called_once_with("file contents")
-        _get_all_htm_files_contents.assert_called_once_with(candidate.mount_points)
+        self.assertEqual(subject, get_target_by_online_id.return_value)
+        read_online_id.assert_called_with(_get_all_htm_files_contents.return_value[0])
+        get_target_by_online_id.assert_called_once_with(slug=online_id.device_slug, target_type=online_id.device_type)
 
-    @mock.patch("mbed_devices._internal.resolve_target.read_product_code")
-    @mock.patch("mbed_devices._internal.resolve_target.read_online_id")
-    @mock.patch("mbed_devices._internal.resolve_target.get_target_by_online_id")
-    def test_will_resolve_targets_using_online_id_when_available(
+    def test_raises_when_target_not_found(
         self, get_target_by_online_id, read_online_id, read_product_code, _get_all_htm_files_contents
     ):
-        _get_all_htm_files_contents.return_value = ["some file contents"]
-        online_id = OnlineId(device_type="hat", device_slug="boat")
-        read_product_code.return_value = None
-        read_online_id.return_value = online_id
+        read_online_id.return_value = OnlineId(device_type="hat", device_slug="boat")
+        get_target_by_online_id.side_effect = UnknownTarget
         candidate = CandidateDeviceFactory()
 
-        subject = _build_target_resolver(candidate)()
+        with self.assertRaises(NoTargetForCandidate):
+            resolve_target(candidate)
 
-        self.assertEqual(subject, get_target_by_online_id.return_value)
-        read_online_id.assert_called_with("some file contents")
-        get_target_by_online_id.assert_called_once_with(slug=online_id.device_slug, target_type=online_id.device_type)
 
-    @mock.patch("mbed_devices._internal.resolve_target.read_product_code")
-    @mock.patch("mbed_devices._internal.resolve_target.read_online_id")
-    @mock.patch("mbed_devices._internal.resolve_target.get_target_by_product_code")
-    def test_will_resolve_targets_by_product_code_from_serial_number_if_htm_file_has_no_information(
+@mock.patch(
+    "mbed_devices._internal.resolve_target._get_all_htm_files_contents",
+    autospec=True,
+    return_value=["who knows file contents"],
+)
+@mock.patch("mbed_devices._internal.resolve_target.read_product_code", autospec=True, return_value=None)
+@mock.patch("mbed_devices._internal.resolve_target.read_online_id", autospec=True, return_value=None)
+@mock.patch("mbed_devices._internal.resolve_target.get_target_by_product_code", autospec=True)
+class TestResolveTargetUsingProductCodeFromSerial(TestCase):
```

Comment:
I'm not convinced we need a test class per resolution case. How about `TestResolveTargets` with a test case for each case.

Reply:
I prefer it as is tbh. Each case has certain mock state that's valid for entire case, so it's easier to group them that way.